src/AppInstallerCommonCore/Public/winget/Yaml.h

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#pragma once
+#include <fstream>
+#include <map>
+#include <memory>
+#include <optional>
+#include <stack>
+#include <string>
+#include <string_view>
+#include <vector>
+
+
+namespace AppInstaller::YAML
+{
+    // A location within the stream.
+    struct Mark
+    {
+        Mark() = default;
+        Mark(size_t l, size_t c) : line(l), column(c) {}
+
+        size_t line = 0;
+        size_t column = 0;
+    };
+
+    // An exception from YAML.
+    struct Exception : public wil::ResultException
+    {
+        // The type of error that occurred.
+        enum class Type
+        {
+            None,
+            Memory,
+            Reader,
+            Scanner,
+            Parser,
+            Composer,
+            Writer,
+            Emitter
+        };
+
+        // Should only be used for Memory.
+        Exception(Type type);
+
+        // Should only be used for Reader.
+        Exception(Type type, const char* problem, size_t offset, int value);
+
+        // Used for Scanner, Parser, and Composer.
+        Exception(Type type, const char* problem, const Mark& problemMark, const char* context = {}, const Mark& contextMark = {});
+
+        // Used for Writer and Emitter.
+        Exception(Type type, const char* problem);
+
+        const char* what() const noexcept override;
+
+    private:
+        std::string m_what;
+    };
+
+    // A YAML node.
+    struct Node
+    {
+        // The node's type.
+        enum class Type
+        {
+            Invalid,
+            None,
+            Scalar,
+            Sequence,
+            Mapping
+        };
+
+        Node() : m_type(Type::Invalid) {}
+        Node(Type type, std::string tag, const Mark& mark);
+
+        // Sets the scalar value of the node.
+        void SetScalar(std::string value);
+
+        // Adds a child node to the sequence.
+        template <typename... Args>
+        Node& AddSequenceNode(Args&&... args)
+        {
+            Require(Type::Sequence);
+            return m_sequence->emplace_back(std::forward<Args>(args)...);
+        }
+
+        // Adds a child node to the mapping.
+        template <typename... Args>
+        Node& AddMappingNode(Node&& key, Args&&... args)
+        {
+            Require(Type::Mapping);
+            return m_mapping->emplace(std::move(key), Node(std::forward<Args>(args)...))->second;
+        }
+
+        bool IsDefined() const { return m_type != Type::Invalid; }
+        bool IsNull() const { return m_type == Type::Invalid || m_type == Type::None || (m_type == Type::Scalar && m_scalar.empty()); }

```

Comment:
>m_type == Type::Invalid [](start = 37, length = 23)

should we allow Invalid as null?

Reply:
This was to match the behavior of yaml-cpp with respect to what we generate from libyaml.

---
In reply to: [468973641](https://github.com/Microsoft/pkg/pull/535#discussion_r468973641) [](ancestors = 468973641)